File: backend/app/models/cluster.py
"""Cluster Pydantic models."""

from typing import Any

from pydantic import BaseModel, Field


class NodeResourceUsage(BaseModel):
    """Per-node resource usage."""
    total: float
    used: float
    free: float
    percentage: float


class NodeResponse(BaseModel):
    """Response model for a Node."""

    name: str
    status: str
    roles: list[str] = Field(default_factory=list)
    version: str | None = None
    os: str | None = None
    cpu: str | None = None
    memory: str | None = None
    internal_ip: str | None = None
    cpu_usage: NodeResourceUsage | None = None
    memory_usage: NodeResourceUsage | None = None


class NodeListResponse(BaseModel):
    """Response model for listing Nodes."""

    items: list[NodeResponse]
    total: int


class ClusterStatusResponse(BaseModel):
    """Response model for cluster status."""

    kubevirt: dict[str, Any]
    cdi: dict[str, Any]
    nodes_count: int
    nodes_ready: int

File: frontend/src/types/cluster.ts
export interface Namespace {
  name: string;
  status: string;
  labels: Record<string, string>;
  created?: string;
}

export interface NamespaceListResponse {
  items: Namespace[];
  total: number;
}

export interface NodeResourceUsage {
  total: number;
  used: number;
  free: number;
  percentage: number;
}

export interface Node {
  name: string;
  status: string;
  roles: string[];
  version?: string;
  os?: string;
  cpu?: string;
  memory?: string;
  internal_ip?: string;
  cpu_usage?: NodeResourceUsage;
  memory_usage?: NodeResourceUsage;
}

export interface NodeListResponse {
  items: Node[];
  total: number;
}

export interface ClusterStatus {
  kubevirt: {
    installed: boolean;
    phase?: string;
    version?: string;
    targetVersion?: string;
    error?: string;
  };
  cdi: {
    installed: boolean;
    phase?: string;
    error?: string;
  };
  nodes_count: number;
  nodes_ready: number;
}

export interface ResourceUsage {
  used: string;
  total: string;
  percentage: number;
}

export interface ClusterSettings {
  cpu_overcommit: number;
}

export interface SchedulableSlot {
  cpu_cores: number;
  memory_gi: number;
  node: string;
}

export interface UserResources {
  vms_total: number;
  vms_running: number;
  cpu: ResourceUsage;
  memory: ResourceUsage;
  storage: ResourceUsage;
  max_schedulable: SchedulableSlot | null;
}

export interface ActivityItem {
  id: string;
  type: string;
  message: string;
  resource_name: string;
  resource_namespace: string;
  timestamp: string;
  icon: string;
}

export interface RecentActivityResponse {
  items: ActivityItem[];
}
